src/util/address.rs

```diff
@@ -467,32 +467,80 @@ mod tests {
 
 use crate::vm::VMBinding;
 
-/// ObjectReference represents address for an object. Compared with Address,
-/// operations allowed on ObjectReference are very limited. No address arithmetics
-/// are allowed for ObjectReference. The idea is from the paper
-/// High-level Low-level Programming (VEE09) and JikesRVM.
+/// `ObjectReference` represents address for an object. Compared with `Address`, operations allowed
+/// on `ObjectReference` are very limited. No address arithmetics are allowed for `ObjectReference`.
+/// The idea is from the paper [Demystifying Magic: High-level Low-level Programming (VEE09)][FBC09]
+/// and [JikesRVM].
+///
+/// In MMTk, `ObjectReference` holds a non-zero address.  It either points into one of MMTk's spaces
+/// or not.
+///
+/// -   When the address is in one of MMTk's spaces, it refers to an object allocated in that space.
+///     In this case, the address must satisfy the following requirements.
+///     -   The address must be within the address range of the object it refers to.  But it is not
+///         necessarily the starting address.
+///     -   The address must be word-aligned.
+/// -   When the address is outside any MMTk space, its semantics is VM-defined.
```

Comment:
I don't think we have well-defined semantics for `ObjectReference` that points outside MMTk spaces. For example, if we access side metadata, we may see unmapped side metadata and we may panic.

Reply:
I did say it is "VM-defined".  But I can emphasize that MMTk won't access side metadata from an `ObjectReference` without knowing it is a valid reference to an MMTk object (or at least in one of MMTk's spaces), and that VM bindings shouldn't access side metadata using `ObjectReference`s that don't point to a valid MMTk object.

Reply:
> (I can emphasize that) MMTk won't access side metadata from an `ObjectReference` without knowing it is a valid reference to an MMTk object (or at least in one of MMTk's spaces)

That's what I was concerned about. I don't think we can guarantee that with the current code base.

This is not the focus of this PR. So I think it is fine to say `ObjectReference` is only for our allocated memory (with VM space being an exception). We don't need to mention `ObjectReference` to non-MMTk memory. A binding may use `ObjectReference` to point to non-MMTk memory -- it is a hack, it may work or  it may not, we don't promise anything.

Reply:
Maybe I can move "ObjectReference to non-MMTk memory" down to the "Notes" section below, and briefly mention `ActivePlan::vm_trace_object` and the fact that this feature is experimental and may not work.  I just feel that the doc comment of `ObjectReference` should not contradict with that of `ActivePlan::vm_trace_object`.

Reply:
Yeah. This sounds good.

Reply:
Done.